File: android/PositiveOnlySocial/app/src/test/java/com/example/positiveonlysocial/util/RelativeTimeTest.kt
package com.example.positiveonlysocial.util

import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.Date

class RelativeTimeTest {

    // Fixed reference point so assertions don't depend on the wall clock.
    private val now = Date(1_780_000_000_000L)

    private fun ago(seconds: Long): Date = Date(now.time - seconds * 1000)

    private val minute = 60L
    private val hour = 60 * minute
    private val day = 24 * hour
    private val week = 7 * day
    private val year = 365 * day

    @Test
    fun collapsesSubMinuteToLessThanOneMin() {
        assertEquals("< 1 min", RelativeTime.format(ago(0), now))
        assertEquals("< 1 min", RelativeTime.format(ago(1), now))
        assertEquals("< 1 min", RelativeTime.format(ago(59), now))
    }

    @Test
    fun futureTimestampsReadLessThanOneMin() {
        assertEquals("< 1 min", RelativeTime.format(Date(now.time + 5000), now))
    }

    @Test
    fun roundsDownToWholeMinutes() {
        assertEquals("1 min", RelativeTime.format(ago(minute), now))
        assertEquals("1 min", RelativeTime.format(ago(minute + 59), now))
        assertEquals("59 min", RelativeTime.format(ago(59 * minute), now))
    }

    @Test
    fun roundsDownToWholeHours() {
        assertEquals("1 hr", RelativeTime.format(ago(hour), now))
        assertEquals("23 hr", RelativeTime.format(ago(23 * hour), now))
    }

    @Test
    fun roundsDownToWholeDaysWithPluralization() {
        assertEquals("1 day", RelativeTime.format(ago(day), now))
        assertEquals("6 days", RelativeTime.format(ago(6 * day), now))
    }

    @Test
    fun roundsDownToWholeWeeksWithPluralization() {
        assertEquals("1 week", RelativeTime.format(ago(week), now))
        assertEquals("8 weeks", RelativeTime.format(ago(8 * week), now))
        // 364 days is still under a year, so it stays in weeks.
        assertEquals("52 weeks", RelativeTime.format(ago(364 * day), now))
    }

    @Test
    fun roundsDownToWholeYearsWithPluralization() {
        assertEquals("1 year", RelativeTime.format(ago(year), now))
        assertEquals("2 years", RelativeTime.format(ago(2 * year + 5 * day), now))
    }
}

File: website/src/utils/relativeTime.test.ts
import { describe, expect, test } from 'vitest'
import { formatRelativeTime } from './relativeTime'

// Fixed reference point so the assertions don't depend on the wall clock.
const now = new Date('2026-06-19T12:00:00Z')

function ago(seconds: number): Date {
  return new Date(now.getTime() - seconds * 1000)
}

const MIN = 60
const HOUR = 60 * MIN
const DAY = 24 * HOUR
const WEEK = 7 * DAY
const YEAR = 365 * DAY

describe('formatRelativeTime', () => {
  test('collapses sub-minute durations to "< 1 min"', () => {
    expect(formatRelativeTime(ago(0), now)).toBe('< 1 min')
    expect(formatRelativeTime(ago(1), now)).toBe('< 1 min')
    expect(formatRelativeTime(ago(59), now)).toBe('< 1 min')
  })

  test('future or equal timestamps read "< 1 min" rather than negative', () => {
    expect(formatRelativeTime(new Date(now.getTime() + 5000), now)).toBe('< 1 min')
  })

  test('rounds down to whole minutes', () => {
    expect(formatRelativeTime(ago(MIN), now)).toBe('1 min')
    expect(formatRelativeTime(ago(MIN + 59), now)).toBe('1 min')
    expect(formatRelativeTime(ago(59 * MIN), now)).toBe('59 min')
  })

  test('rounds down to whole hours', () => {
    expect(formatRelativeTime(ago(HOUR), now)).toBe('1 hr')
    expect(formatRelativeTime(ago(23 * HOUR), now)).toBe('23 hr')
  })

  test('rounds down to whole days with pluralization', () => {
    expect(formatRelativeTime(ago(DAY), now)).toBe('1 day')
    expect(formatRelativeTime(ago(6 * DAY), now)).toBe('6 days')
  })

  test('rounds down to whole weeks with pluralization', () => {
    expect(formatRelativeTime(ago(WEEK), now)).toBe('1 week')
    expect(formatRelativeTime(ago(8 * WEEK), now)).toBe('8 weeks')
    // 364 days is still under a year, so it stays in weeks.
    expect(formatRelativeTime(ago(364 * DAY), now)).toBe('52 weeks')
  })

  test('rounds down to whole years with pluralization', () => {
    expect(formatRelativeTime(ago(YEAR), now)).toBe('1 year')
    expect(formatRelativeTime(ago(2 * YEAR + 5 * DAY), now)).toBe('2 years')
  })

  test('accepts ISO strings and returns "" for unparseable input', () => {
    expect(formatRelativeTime(ago(5 * MIN).toISOString(), now)).toBe('5 min')
    expect(formatRelativeTime('not a date', now)).toBe('')
  })
})
